### DISCUSSION TITLE
Compile EdgeDB queries & Rust

### DISCUSSION THREAD

**Discussion by pksunkara** [geldata/gel#4017] [Q&A]
Hello,

I am looking to use EdgeDB with [rust](https://rust-lang.com) and since it's a compiled language, I am wondering whether I can translate EdgeQL into SQL during compilation itself and thus not need the EdgeDB server in production.

I haven't found enough documentation and am thus working with a few assumptions here:

1. EdgeDB doesn't actually interact with Postgres other than to retrieve and update data.
2. EdgeDB can be spun up as a pure query translator. It reads the schema and query and translates it to SQL.

So, my questions for the developer: How valid are my assumptions? If not, is it feasible?

**Comment by nsidnev** [comment]
So, I'm not an EdgeDB developer, but here's what I know about some of its internals regarding your assumptions:

1. EdgeDB is very closely coupled with PostgreSQL. When EdgeDB is bootstraped, some internal tables with many functions are created. These functions are used by EdgeDB when converting EdgeQL to SQL, and without them the generated SQL won't work.
Also, when EdgeDB works with the schema, it does not display names one-to-one, as defined in SDL, but "mangles" them into internal identifiers (like `edgedbstd."c4a26e80-f541-11ec-ad16-4d10b5fd181c_t"`). You can try to check them by connecting to a PostgreSQL server, although I am not sure that anyone other than the compiler itself can figure out what is behind these identifiers :)
And after each operation, EdgeDB translates the results from PostgreSQL (including errors that may contain these meaningless identifiers in messages) into a form that can be useful to users. In case of successful operation in objects, scalars, etc. In case of errors into EdgeDB errors with human-readable messages.

2. This follows from the first answer. This is not possible by default, although *theoretically* you can do it. You can run EdgeDB at compile time, tell it to generate SQL for EdgeQL, save it somewhere in a file, and then use the resulting query with the PostgreSQL driver. But, as already mentioned, these queries will be quite difficult to read and understand, and if an error occurs (for example, you try to insert a record into a non-existent table), the error message won't be clear either.

As to the last question about whether such options are possible, I'm not so sure. The developers mentioned that in the future they plan to make it possible to inspect EdgeDB via SQL queries, but I think it will be a read-only interface and will be used mainly for collecting information, not for modifying the database.

**Reply by tailhook** [reply]
> You can run EdgeDB at compile time, tell it to generate SQL for EdgeQL, save it somewhere in a file, and then use the resulting query with the PostgreSQL driver.

Apart from downsides mentioned by @nsidnev, I have to also add that you will have to perfectly sync those compiled queries when schema changes (basically stop the app, upgrade schema, spin up the new version of the app). Otherwise, constraints enforced by EdgeDB might not work and you can get an inconsistent database.

So generally we don't support this kind of usage and may develop features that will make this kind of usage even harder.

**Reply by pksunkara** [reply]
> I have to also add that you will have to perfectly sync those compiled queries when schema changes (basically stop the app, upgrade schema, spin up the new version of the app). Otherwise, constraints enforced by EdgeDB might not work and you can get an inconsistent database.

That should be okay with a simple tool.

> So generally we don't support this kind of usage and may develop features that will make this kind of usage even harder.

Hmm... the direction of the tool is not for me then. Thanks for your time.

**Reply by avin-kavish** [reply]
Why do you think compiling queries is going to add value to your use case?
